aisdc/generate_report.py

```diff
@@ -0,0 +1,312 @@
+"""
+Generate report for TREs from JSON file
+"""
+
+import json
+import os
+import pprint
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class AnalysisModule:
+    """
+    Wrapper module for metrics analysis modules
+    """
+
+    def process_dict(self):
+        """
+        Function that produces a risk summary output based on analysis in this module
+        """
+        raise NotImplementedError()
+
+    def __str__(self):
+        raise NotImplementedError()
+
+
+class FinalRecommendationModule(
+    AnalysisModule
+):  # pylint: disable=too-many-instance-attributes
+    """
+    Module that generates the first layer of a recommendation report
+    """
+
+    def __init__(self, report: dict):
+        self.P_VAL_THRESH = 0.05
+        self.MEAN_AUC_THRESH = 0.65
+
+        self.SVM_WEIGHTING_SCORE = 5
+        self.MIN_SAMPLES_LEAF_SCORE = 3
+        self.STATISTICALLY_SIGNIFICANT_SCORE = 2
+        self.MEAN_AUC_SCORE = 4
+
+        self.report = report
+
+        self.scores = []
+        self.reasons = []
+
+    def _is_svm(self, svm_weighing_score):
+        if "model" in self.report:
+            if self.report["model"] == "SVC":
+                self.scores.append(svm_weighing_score)
+                self.reasons.append("Model is SVM (supersedes all other tests)")
+                return True
+        return False
+
+    def _rf_min_samples_leaf(self, min_samples_leaf_score):
+        if "model_params" in self.report:
+            if "min_samples_leaf" in self.report["model_params"]:
+                min_samples_leaf = self.report["model_params"]["min_samples_leaf"]
+                if min_samples_leaf < 5:
+                    self.scores.append(min_samples_leaf_score)
+                    self.reasons.append("Min samples per leaf < 5")
+
+    def _statistically_significant_auc(
+        self, p_val_thresh, mean_auc_thresh, stat_sig_score, mean_auc_score
+    ):
+        stat_sig_auc = []
+        if "attack_experiment_logger" in self.report:
+            for i in self.report["attack_experiment_logger"]["attack_instance_logger"]:
+                instance = self.report["attack_experiment_logger"][
+                    "attack_instance_logger"
+                ][i]["metrics_array"]
+                if instance["P_HIGHER_AUC"] < p_val_thresh:
+                    stat_sig_auc.append(instance["AUC"])
+
+            n_instances = len(
+                self.report["attack_experiment_logger"]["attack_instance_logger"]
+            )
+            if (
+                len(stat_sig_auc) / n_instances > 0.1
+            ):  # > 10% of AUC are statistically significant
```

Comment:
Tiny thing here -- no multple test correction happening here, so we'd expect 1 in 20 attacks to give a significant result by chance. If 1 in 20 happen by chance, then exceeding the 0.1 threshold is probably not that unlikely? Don't need to change this now, but it might give confusing output if the module below says nothing is significant after correcting for multiple testing and this says that there's a problem?

Reply:
@yolaj-nhs @simonrnss @rpreen 
Something odd has suddenly started happening with what is obviously a new version of tensorflow_privacy.
I will update safekearas which should fix the lining and test issues